### PR TITLE
New C++ Api: Second and last batch of API guards.

### PR DESCRIPTION
**Pull request by aniemetz**
This adds the remaining API guards in the Solver object (incl. unit tests).

### CODE REVIEW

File: test/unit/api/solver_black.h [hunk 1]
```diff
@@ -1132,6 +1231,150 @@ void SolverBlack::testGetOp()
...
+  d_solver->assertFormula(d_solver->mkTerm(LEQ, zero, f_x));
+  d_solver->assertFormula(d_solver->mkTerm(LEQ, zero, f_y));
+  d_solver->assertFormula(d_solver->mkTerm(LEQ, sum, one));
+  d_solver->assertFormula(p_0.notTerm());
+  d_solver->assertFormula(p_f_y);
+  TS_ASSERT(d_solver->checkSat().isSat());
+  TS_ASSERT_THROWS_NOTHING(d_solver->getValue(x));
+  TS_ASSERT_THROWS_NOTHING(d_solver->getValue(y));
+  TS_ASSERT_THROWS_NOTHING(d_solver->getValue(sum));
+  TS_ASSERT_THROWS_NOTHING(d_solver->getValue(p_f_y));
```
Comment (4tXJ7f) [inline test/unit/api/solver_black.h:1372]: What happens if we ask for a term that is not in the assertions at all? E.g. a fresh variable?

File: test/unit/api/solver_black.h [hunk 2]
```diff
@@ -1132,6 +1231,150 @@ void SolverBlack::testGetOp()
...
+  d_solver->assertFormula(d_solver->mkTerm(GT, zero, f_y));
+  d_solver->assertFormula(d_solver->mkTerm(GT, sum, one));
+  d_solver->assertFormula(p_0);
+  d_solver->assertFormula(p_f_y.notTerm());
+  TS_ASSERT(d_solver->checkSat().isUnsat());
+
+  TS_ASSERT_THROWS_NOTHING(unsat_core = d_solver->getUnsatCore());
+
+  d_solver->resetAssertions();
+  for (auto& t : unsat_core)
```
Comment (4tXJ7f) [inline test/unit/api/solver_black.h:1315]:
```suggestion
  for (const auto& t : unsat_core)
```

File: test/unit/api/solver_black.h [hunk 3]
```diff
@@ -1068,6 +1125,42 @@ void SolverBlack::testDefineFunsRec()
...
+      CVC4ApiException&);
+  TS_ASSERT_THROWS(
+      slv.defineFunsRec({f12, f22}, {{b12, b112}, {b4}}, {v12, v22}),
+      CVC4ApiException&);
+  TS_ASSERT_THROWS(
+      slv.defineFunsRec({f12, f22}, {{b12, b112}, {b42}}, {v12, v2}),
+      CVC4ApiException&);
+  TS_ASSERT_THROWS(
+      slv.defineFunsRec({f12, f22}, {{b12, b112}, {b42}}, {v1, v22}),
+      CVC4ApiException&);
```
Comment (4tXJ7f) [inline test/unit/api/solver_black.h:1163]:
```suggestion
  TS_ASSERT_THROWS(
      slv.defineFunsRec({f12, f22}, {{b12, b112}, {b42}}, {v1, v22}),
      CVC4ApiException&);
  TS_ASSERT_THROWS(
      slv.defineFunsRec({f12, f22}, {{b12, b112}, {b42}}, {v12, v2}),
      CVC4ApiException&);
```

(just to swapping the order to match the rest :) )

File: src/api/cvc4cpp.cpp [hunk 4]
```diff
@@ -4583,29 +4654,28 @@ void Solver::push(uint32_t nscopes) const
...
-  {
-    CVC4::LogicInfo logic_info(logic);
-    d_smtEngine->setLogic(logic_info);
-  }
-  catch (CVC4::IllegalArgumentException& e)
-  {
-    throw CVC4ApiException(e.getMessage());
-  }
+  CVC4::LogicInfo logic_info(logic);
+  d_smtEngine->setLogic(logic_info);
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.cpp:4670]: Don't we need `CVC4_API_SOLVER_TRY_CATCH_BEGIN` and `CVC4_API_SOLVER_TRY_CATCH_END` in this method?

File: src/api/cvc4cpp.cpp [hunk 5]
```diff
@@ -4432,51 +4459,70 @@ std::vector<Term> Solver::getAssertions(void) const
...
  *  ( get-assignment )
  */
 std::vector<std::pair<Term, Term>> Solver::getAssignment(void) const
 {
-  // CHECK: produce-models set
-  // CHECK: result sat
+  CVC4_API_SOLVER_TRY_CATCH_BEGIN;
+  CVC4::ExprManagerScope exmgrs(*(d_exprMgr.get()));
+  CVC4_API_CHECK(CVC4::options::produceAssignments())
+      << "Cannot get assignment unless model generation is enabled "
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.cpp:4473]:
```suggestion
      << "Cannot get assignment unless assignment production is enabled "
```

File: src/api/cvc4cpp.cpp (hunk 4 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp:4670]: No, not in the helper. It's guarded in the function that calls it.

Comment (aniemetz) [reply src/api/cvc4cpp.cpp:4670]: I removed the helper for now, it was only used once.

File: test/unit/api/solver_black.h (hunk 1 above)
Comment (aniemetz) [reply test/unit/api/solver_black.h:1372]: Nothing (it doesn't throw) -- it generates a value for the given term, as expected. I added it.

Comment (aniemetz) [reply test/unit/api/solver_black.h:1372]: Nothing (it doesn't throw) -- it generates a value for the term. I added it.

File: src/api/cvc4cpp.cpp (hunk 4 above)
Comment (4tXJ7f) [reply src/api/cvc4cpp.cpp]: Ah I missed the fact that this was a helper, my bad (I just saw that we previously were catching the exception).

**Pull request merged**